File: r-package/R/read_health_region.R
#' Download spatial data of Brazilian health regions and health macro regions
#'
#' @description
#' Health regions are used to guide the the regional and state planning of health
#' services. Macro health regions, in particular, are used to guide the planning
#' of high complexity #' health services. These services involve larger economics
#' of scale and are concentrated in few municipalities because they are generally
#' more technology intensive, costly and face shortages of specialized
#' professionals. A macro region comprises one or more health regions.
#'
#' @template year
#' @template code_state
#' @param geometry_level String. Spatial level of the output geometries. Use
#'        `"municipality"` to return municipal geometries (default), `"micro"`
#'        to aggregate geometries by health region, or `"macro"` to aggregate
#'        geometries by health macroregion.
#' @param macro The argument `macro` has been deprecated.
#' @template simplified
#' @template output
#' @template showProgress
#' @template cache
#' @template verbose
#'
#' @return An `"sf" "data.frame"` OR an `ArrowObject`
#'
#' @export
#'
#' @examplesIf identical(tolower(Sys.getenv("NOT_CRAN")), "true")
#' # Read municipalities with info on health regions
#' health_muni <- read_health_region(year = 2024)
#'
#' # Read the geometries of micro regions
#' health_micro <- read_health_region(
#'   year = 2024,
#'   geometry_level = "micro"
#'   )
#'
#' # Read the geometries of macro regions
#' health_macro <- read_health_region(
#'   year = 2024,
#'   geometry_level = "macro"
#' )
#'
read_health_region <- function(year,
                               code_state = "all",
                               geometry_level = "municipality",
                               macro = NULL,
                               simplified = TRUE,
                               output = "sf",
                               showProgress = TRUE,
                               cache = TRUE,
                               verbose = TRUE){


  if (!is.null(macro)) {
    cli::cli_abort(c(
      "Argument deprecated",
      "x" = "The `macro` argument is deprecated. Use `geometry_level` instead."
    ))
  }


  # check input
  allowed <- c("municipality", "micro", "macro")
  if (!all(geometry_level %in% allowed)) {
    cli::cli_abort(c(
      "`geometry_level` must be one of: {.val {allowed}}.",
      "x" = "Invalid value{?s}: {.val {setdiff(geometry_level, allowed)}}."
    ))
  }

  # Get metadata
  temp_meta <- select_metadata(
    geography = "healthregions",
    year = year,
    simplified = simplified,
    verbose = verbose
  )

  # check if download failed
  if (is.null(temp_meta)) { return(invisible(NULL)) }

  # download file and open arrow dataset
  temp_arrw <- download_parquet(
    filename_to_download = temp_meta$file_name,
    showProgress = showProgress,
    cache = cache
  )

  # check if download failed
  if (is.null(temp_arrw)) { return(invisible(NULL)) }

  # FILTER
  temp_arrw <- filter_arrw(temp_arrw, code = code_state)

  # geometry_level
  if (geometry_level=="municipality") {

    # convert to sf
    temp <- convert_output(temp_arrw, output)

    return(temp)
  }

  # if micro or macro, perform aggregation
    # convert to sf
    temp <- convert_output(temp_arrw, "sf")

    all_cols <- names(temp)

    if(geometry_level=="micro"){
      group_cols <- all_cols[!grepl('geometry|code_muni|name_muni|code_health_macroregion|name_health_macroregion', all_cols)]
    } else {
      group_cols <- all_cols[!grepl('geometry|code_muni|name_muni|code_health_region|name_health_region', all_cols)]
    }

    temp <- duckspatial::ddbs_union_agg(
      x = temp,
      by = group_cols
      ) |>
      duckspatial::ddbs_collect() |>
      sfheaders::sf_remove_holes()

  # convert to sf
    temp <- convert_output(temp, output)

  return(temp)

}

File: python-package/geobr/read_health_region.py
import warnings

from geobr.utils import read_geobr_v2
from geobr._output import convert_output
from geobr._duckdb_backend import duckdb_connection


def read_health_region(
    year: int,
    code_state: str = "all",
    geometry_level: str = "municipality",
    macro=None,
    simplified: bool = True,
    verbose: bool = False,
    output: str = "gpd",
    show_progress: bool = True,
    cache: bool = True,
):
    """Download Brazilian health region data.

    Parameters
    ----------
    year : int
        Year of the data.
    code_state : str or int
        State abbrev, two-digit code, or ``"all"``.
    geometry_level : str
        ``"municipality"`` (default), ``"micro"``, or ``"macro"``.
    macro : bool, optional
        Deprecated. Use ``geometry_level`` instead.
    simplified, verbose, output, show_progress, cache
        Standard geobr options.
    """
    if macro is not None:
        warnings.warn(
            "The `macro` argument is deprecated. Use `geometry_level` instead.",
            DeprecationWarning,
            stacklevel=2,
        )
        geometry_level = "macro" if macro else "municipality"

    allowed = ("municipality", "micro", "macro")
    if geometry_level not in allowed:
        raise ValueError(
            f"`geometry_level` must be one of: {list(allowed)}. Got: {geometry_level!r}"
        )

    relation = read_geobr_v2(
        "healthregions",
        year,
        code=code_state,
        simplified=simplified,
        output="duckdb",
        show_progress=show_progress,
        cache=cache,
        verbose=verbose,
    )

    conn = duckdb_connection()

    if geometry_level == "municipality":
        return convert_output(relation, output, conn)

    all_cols = relation.columns

    if geometry_level == "micro":
        group_cols = [
            c
            for c in all_cols
            if c != "geometry"
            and c not in ("code_muni", "name_muni", "code_health_macroregion", "name_health_macroregion")
        ]
    else:
        group_cols = [
            c
            for c in all_cols
            if c != "geometry"
            and c not in ("code_muni", "name_muni", "code_health_region", "name_health_region")
        ]

    group_cols_str = ", ".join(group_cols)

    # Aggregate results and remove holes
    query = f"""
        WITH aggregated AS (
            -- perform the standard union aggregation
            SELECT 
                {group_cols_str},
                ST_Union_Agg(geometry) AS geom
            FROM relation
            GROUP BY {group_cols_str}
        ),
        unwrapped_polygons AS (
            -- flatten multipolygons into separate rows of simple polygons
            SELECT 
                {group_cols_str},
                (UNNEST(ST_Dump(geom))).geom AS single_geom
            FROM aggregated
        ),
        holes_removed AS (
            -- remove holes from the simple polygons using the outer ring
            SELECT 
                {group_cols_str},
                ST_MakePolygon(ST_ExteriorRing(single_geom)) AS clean_geom
            FROM unwrapped_polygons
        )
        -- recollect the cleaned parts back into the final shapes
        SELECT 
            {group_cols_str},
            ST_Union_Agg(clean_geom) AS geometry
        FROM holes_removed
        GROUP BY {group_cols_str};
        """

    relation = conn.sql(query)

    return convert_output(relation, output, conn)
